docs/codeql/ql-language-reference/aliases.rst

```diff
@@ -123,4 +123,18 @@ You could give the predicate a more descriptive name as follows:
 
 .. code-block:: ql
 
-    predicate lessThanTen = isSmall/1;
\ No newline at end of file
+    predicate lessThanTen = isSmall/1;
+
+.. _weak_strong_aliases:
+
+Strong and weak aliases
+=======================
+
+Every alias is either **strong** or **weak**.
+An alias is **strong** if and only if it is a :ref:`type alias <type-aliases>` with :ref:`annotation <annotations>`
+``final``.
+During :ref:`name resolution <name-resolution>`, ambiguity between **weak** aliases of the same target is allowed,
+but ambiguity between distinct **strong** aliases or entities is invalid QL.
```

Comment:
I think we need to distinguish between alias definition and alias binding here. We cannot really say "an alias binding is strong" or "an alias binding is weak". However, we can say "an alias definition is strong" or and "alias definition is weak". In your first sentence, you implicitly say two things:
- an alias of an alias of an entity is an alias of that entity and
- being weak or strong is a property of an alias in relation to the entity.

You say `MyIntAlias` should be a strong alias. Otherwise, why would there be an ambiguity error? But the alias definition `class MyIntAlias = MyInt1` _is_ a weak alias definition. `MyIntAlias` is a weak alias for `MyInt1`.
What you mean is that `MyIntAlias` and its underlying entity are not equal `equal modulo weak aliasing`, but that level of detail is for the language specification imho.

I've pushed another commit that attempts to make these distinctions.

Reply:
That is fair. I see that the alias language reference page only talks about alias to entities as well. That is probably a sensible simplification and then the language specification can go into full details about nested aliases.